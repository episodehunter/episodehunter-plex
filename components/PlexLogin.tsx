import * as React from 'react';

type PlexLoginProps = {username: string, password: string, onSucess: (username, token) => void, onCancel: () => void};
type PlexLoginState = {username?: string, password?: string, token?: string, loading?: boolean};

class FlavorForm extends React.Component<PlexLoginProps, PlexLoginState> {

  constructor(props: PlexLoginProps) {
    super(props);
    this.state = {
      username: props.username,
      password: props.password,
      loading: false
    };
  }

  login() {
    console.log('Login');
  }

  chnage(prop: string, event) {
    this.setState({[prop]: event.target.value});
  }

  render() {
    return (
      <div>
        <div className='from-row'>
          <label>Plex username:</label>
          <input type="text" value={this.state.username} onChange={e => this.chnage('username', e)} />
        </div>
        <div className="from-row">
          <label>Plex password:</label>
          <input type="password" value={this.state.password} onChange={e => this.chnage('password', e)} />
        </div>
        <button onClick={() => this.login()}>Login</button>
        <p onClick={this.props.onCancel}>Cancel</p>
      </div>
    );
  }
}
