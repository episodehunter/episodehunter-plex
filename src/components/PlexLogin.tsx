import * as React from 'react';
import { Subject } from 'rxjs/Subject';
import { Observable } from 'rxjs/Observable';
import { Subscription } from 'rxjs/Subscription';
import { ajax } from 'rxjs/observable/dom/ajax';
import { Overlay } from './Overlay';

type PlexLoginProps = {username: string, token: string, onSucess: (username, token) => void, onCancel: () => void};
type PlexLoginState = {username?: string, password?: string, token?: string, loading?: boolean, errorMsg?: string};

export class PlexLogin extends React.Component<PlexLoginProps, PlexLoginState> {
  loginClick$ = new Subject();
  loginClickSubscription: Subscription;

  constructor(props: PlexLoginProps) {
    super(props);
    this.state = {
      username: props.username,
      password: '',
      token: props.token,
      loading: false,
      errorMsg: ''
    };

    this.loginClickSubscription = this.loginClick$
      .do(() => this.setState({loading: true}))
      .map(() => `user[login]=${encodeURIComponent(this.state.username)}&user[password]=${encodeURIComponent(this.state.password)}`)
      .switchMap(formData => {
        return ajax
          .post(
            'https://plex.tv/users/sign_in.json',
            formData,
            { 'Content-Type': 'application/x-www-form-urlencoded', 'X-Plex-Client-Identifier': 'episodehunter', 'X-Plex-Product': 'Episodehunter', 'X-Plex-Version': '1.0.0' }
          )
          .do(response => this.props.onSucess(this.state.username, response.response.user.authToken))
          .catch(reason => {
            console.log(reason);
            this.setState({
              loading: false,
              errorMsg: 'Something went wrong'
            });
            return Observable.of(null);
          });
      })
      .subscribe();
  }

  componentWillUnmount() {
    this.loginClickSubscription.unsubscribe();
  }

  change(prop: keyof PlexLoginState, event) {
    this.setState({[prop]: event.target.value});
  }

  render() {
    return (
      <div>
        {this.state.loading && <Overlay />}
        <div className='from-row'>
          <label>Plex username:</label>
          <input type="text" value={this.state.username} onChange={e => this.change('username', e)} />
        </div>
        <div className="from-row">
          <label>Plex password:</label>
          <input type="password" value={this.state.password} onChange={e => this.change('password', e)} />
        </div>
        <div style={{color: 'red'}}>
         {this.state.errorMsg}
        </div>
        <button onClick={() => this.loginClick$.next()}>Request token</button>
        <p onClick={this.props.onCancel}>Cancel</p>
      </div>
    );
  }
}
