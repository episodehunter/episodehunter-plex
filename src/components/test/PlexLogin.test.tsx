import * as React from 'react';
import * as renderer from 'react-test-renderer';
import { PlexLogin } from '../PlexLogin';

test('that PlexLogin renders', () => {
  const fun = () => {};
  const tree = renderer.create(
    <PlexLogin onCancel={fun} onSucess={fun} token="" username="" />
  ).toJSON();
  expect(tree).toMatchSnapshot();
});
