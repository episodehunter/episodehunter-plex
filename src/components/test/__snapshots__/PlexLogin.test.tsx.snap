// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`that PlexLogin renders 1`] = `
<div
  style={
    Object {
      "width": "100%",
    }
  }
>
  <div
    className="from-row"
  >
    <label>
      Plex username:
    </label>
    <input
      onChange={[Function]}
      type="text"
      value=""
    />
  </div>
  <div
    className="from-row"
  >
    <label>
      Plex password:
    </label>
    <input
      onChange={[Function]}
      type="password"
      value=""
    />
  </div>
  <div
    style={
      Object {
        "color": "red",
      }
    }
  >
    
  </div>
  <div
    style={
      Object {
        "display": "flex",
        "justifyContent": "space-between",
      }
    }
  >
    <button
      onClick={[Function]}
    >
      Request token
    </button>
    <button
      className="cancel"
      onClick={[Function]}
    >
      Cancel
    </button>
  </div>
</div>
`;
